src/graphic/Fast3D/gfx_opengl.cpp

```diff
@@ -447,26 +459,44 @@ static struct ShaderProgram* gfx_opengl_create_and_load_new_shader(uint64_t shad
             } else {
                 if (s && t) {
                     fs_len += sprintf(fs_buf + fs_len,
-                                      "vec4 texVal%d = hookTexture2D(uTex%d, clamp(vTexCoord%d, 0.5 / texSize%d, "
-                                      "vec2(vTexClampS%d, vTexClampT%d)), texSize%d);\n",
-                                      i, i, i, i, i, i, i);
+                                      "vec2 vTexCoordAdj%d = clamp(vTexCoord%d, 0.5 / texSize%d, "
+                                      "vec2(vTexClampS%d, vTexClampT%d));\n",
+                                      i, i, i, i, i);
                 } else if (s) {
                     fs_len += sprintf(fs_buf + fs_len,
-                                      "vec4 texVal%d = hookTexture2D(uTex%d, vec2(clamp(vTexCoord%d.s, 0.5 / "
-                                      "texSize%d.s, vTexClampS%d), vTexCoord%d.t), texSize%d);\n",
-                                      i, i, i, i, i, i, i);
+                                      "vec2 vTexCoordAdj%d = vec2(clamp(vTexCoord%d.s, 0.5 / "
+                                      "texSize%d.s, vTexClampS%d), vTexCoord%d.t);\n",
+                                      i, i, i, i, i);
                 } else {
                     fs_len += sprintf(fs_buf + fs_len,
-                                      "vec4 texVal%d = hookTexture2D(uTex%d, vec2(vTexCoord%d.s, clamp(vTexCoord%d.t, "
-                                      "0.5 / texSize%d.t, vTexClampT%d)), texSize%d);\n",
-                                      i, i, i, i, i, i, i);
+                                      "vec2 vTexCoordAdj%d = vec2(vTexCoord%d.s, clamp(vTexCoord%d.t, "
+                                      "0.5 / texSize%d.t, vTexClampT%d));\n",
+                                      i, i, i, i, i);
                 }
             }
+
+            fs_len += sprintf(fs_buf + fs_len, "vec4 texVal%d = hookTexture2D(uTex%d, vTexCoordAdj%d, texSize%d);\n", i,
+                              i, i, i);
+            if (cc_features.used_masks[i]) {
+                fs_len += sprintf(fs_buf + fs_len, "vec2 maskSize%d = textureSize(uTexMask%d, 0);\n", i, i);
+
+                fs_len +=
+                    sprintf(fs_buf + fs_len,
+                            "vec4 maskVal%d = hookTexture2D(uTexMask%d, vTexCoordAdj%d, maskSize%d);\n", i, i, i, i);
+                if (cc_features.used_blend[i]) {
+                    fs_len += sprintf(fs_buf + fs_len,
+                                      "vec4 blendVal%d = hookTexture2D(uTexBlend%d, vTexCoordAdj%d, texSize%d);\n", i,
+                                      i, i, i);
+                } else {
+                    fs_len += sprintf(fs_buf + fs_len, "vec4 blendVal%d = vec4(0, 0, 0, 0);\n", i);
+                }
+
+                fs_len += sprintf(fs_buf + fs_len, "texVal%d = mix(texVal%d, blendVal%d, maskVal%d.a);\n", i, i, i, i);
+            }
```

Comment:
This block of code had been reworked to more closely match DirectX, in that it calculates texture coordinates based of wrap/repeat settings, then has a common texture sampling point.